File: app/src/main/java/com/mrboomdev/awery/extensions/support/anilist/data/AnilistMedia.java
package com.mrboomdev.awery.extensions.support.anilist.data;

import static com.mrboomdev.awery.data.Constants.ANILIST_CATALOG_ITEM_ID_PREFIX;
import static com.mrboomdev.awery.util.NiceUtils.stream;

import com.mrboomdev.awery.extensions.data.CatalogMedia;
import com.squareup.moshi.Json;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class AnilistMedia {
	public List<AnilistEpisode> streamingEpisodes;
	public List<String> genres;
	public List<AnilistTag> tags;
	public String description, bannerImage, countryOfOrigin;
	public CoverImage coverImage;
	public MediaTitle title;
	public MediaFormat format;
	public Integer id, duration, episodes, averageScore;
	public MediaDate startDate;
	public boolean isAdult;
	public MediaStatus status;

	public enum MediaFormat {
		TV, TV_SHORT, MOVIE, SPECIAL, OVA, ONA, MUSIC, ONE_SHOT, NOVEL, MANGA
	}

	public static class MediaDate {
		public Integer year, month, day;
	}

	public enum MediaStatus {
		FINISHED, RELEASING, NOT_YET_RELEASED, CANCELLED, HIATUS
	}

	public enum MediaType {
		ANIME, MANGA
	}

	public static class MediaTitle {
		@Json(name = "native")
		public String nativeTitle;
		public String romaji, english;
	}

	public static class CoverImage {
		public String extraLarge, large, color, medium;
	}

	public CatalogMedia toCatalogMedia() {
		var media = new CatalogMedia(ANILIST_CATALOG_ITEM_ID_PREFIX + id);
		media.description = description;
		media.setId("anilist", String.valueOf(id));
		media.banner = bannerImage;
		media.duration = duration;
		media.episodesCount = episodes;
		media.country = countryOfOrigin;
		media.averageScore = (averageScore != null) ? (averageScore / 10f) : null;
		media.genres = new ArrayList<>(genres);
		media.ageRating = isAdult ? "R" : null;

		if(startDate != null && startDate.year != null) {
			media.releaseDate = Calendar.getInstance();

			if(startDate.year != null) media.releaseDate.set(Calendar.YEAR, startDate.year);
			if(startDate.month != null) media.releaseDate.set(Calendar.MONTH, startDate.month - 1);
			if(startDate.day != null) media.releaseDate.set(Calendar.DAY_OF_MONTH, startDate.day);
		}

		if(title.english != null) media.titles.add(title.english);
		if(title.romaji != null) media.titles.add(title.romaji);
		if(title.nativeTitle != null) media.titles.add(title.nativeTitle);

		media.tags = stream(tags)
				.map(AnilistTag::toCatalogTag)
				.toList();

		media.status = switch(status) {
			case RELEASING -> CatalogMedia.MediaStatus.ONGOING;
			case FINISHED -> CatalogMedia.MediaStatus.COMPLETED;
			case HIATUS -> CatalogMedia.MediaStatus.PAUSED;
			case CANCELLED -> CatalogMedia.MediaStatus.CANCELLED;
			default -> CatalogMedia.MediaStatus.COMING_SOON;
		};

		media.poster.medium = coverImage.medium;
		media.poster.large = coverImage.large;
		media.poster.extraLarge = coverImage.extraLarge;

		return media;
	}
}

File: extension/bundled/src/commonMain/kotlin/com/mrboomdev/awery/extension/bundled/anilist/entity/AnilistMedia.kt
package com.mrboomdev.awery.extension.bundled.anilist.entity

import com.mrboomdev.awery.extension.sdk.Media
import kotlinx.serialization.Serializable

@Serializable
class AnilistMedia(
    val id: Int,
    val type: Type,
    val title: Title,
    val description: String?,
    val bannerImage: String?,
    val startDate: FuzzyDate,
    val isAdult: Boolean,
    val siteUrl: String,
    val coverImage: CoverImage,
    val genres: List<String>,
    val tags: List<Tag>,
    val episodes: Int?,
    val chapters: Int?,
    val countryOfOrigin: String?
) {
    @Serializable
    data class Title(
        val english: String?,
        val romaji: String?,
        val native: String?
    )
    
    @Serializable
    data class CoverImage(
        val extraLarge: String?,
        val large: String?,
        val medium: String?
    )
    
    @Serializable
    data class Tag(
        val name: String,
        val isAdult: Boolean
    )
    
    @Serializable
    enum class Type {
        ANIME, MANGA
    }
    
    @Serializable
    enum class Status {
        FINISHED, RELEASING, NOT_YET_RELEASED, CANCELLED, HIATUS
    }
    
    @Serializable
    enum class Sort {
        START_DATE, 
        START_DATE_DESC,
        SCORE, 
        SCORE_DESC,
        POPULARITY,
        POPULARITY_DESC,
        TRENDING,
        TRENDING_DESC,
        UPDATED_AT,
        UPDATED_AT_DESC,
        SEARCH_MATCH,
        FAVOURITES,
        FAVOURITES_DESC
    }
}

fun AnilistMedia.toMedia(): Media {
    val primaryTitle = title.english ?: title.romaji ?: title.native ?: "No title"
    
    return Media(
        id = id.toString(),
        title = primaryTitle,
        description = description,
        banner = bannerImage,
        // Medium has very bad quality so we do prefer a "large" one
        poster = coverImage.let { it.large ?: it.extraLarge ?: it.medium },
        largePoster = coverImage.let { it.extraLarge ?: it.large ?: it.medium },
        ageRating = if(isAdult) "NSFW" else null,
        releaseDate = startDate.toMillis(),
        url = siteUrl,
        episodes = episodes ?: chapters,
        country = countryOfOrigin,
        tags = genres + tags.map { it.name },

        type = when(type) {
            AnilistMedia.Type.ANIME -> Media.Type.WATCHABLE
            AnilistMedia.Type.MANGA -> Media.Type.READABLE
        },

        alternativeTitles = listOfNotNull(
            title.english,
            title.romaji,
            title.native
        ).filterNot { it == primaryTitle }
    )
}